File: backend/app/schemas/io_tree.py
from pydantic import BaseModel, Field


class IOTreeChannelDevice(BaseModel):
    equipment_in_operation_id: int
    equipment_name: str
    manufacturer_name: str | None = None
    nomenclature_number: str | None = None
    article: str | None = None
    ai_count: int
    di_count: int
    ao_count: int
    do_count: int
    signals_total: int


class IOTreeCabinet(BaseModel):
    id: int
    name: str
    factory_number: str | None = None
    inventory_number: str | None = None
    channel_devices: list[IOTreeChannelDevice] = Field(default_factory=list)


class IOTreeLocation(BaseModel):
    id: int
    name: str
    children: list["IOTreeLocation"] = Field(default_factory=list)
    cabinets: list[IOTreeCabinet] = Field(default_factory=list)


class IOTreeResponse(BaseModel):
    locations: list[IOTreeLocation] = Field(default_factory=list)


IOTreeLocation.model_rebuild()

File: frontend/src/api/ioTree.ts
import { apiFetch } from "./client";

export type IOTreeChannelDevice = {
  equipment_in_operation_id: number;
  equipment_name: string;
  manufacturer_name?: string | null;
  nomenclature_number?: string | null;
  article?: string | null;
  ai_count: number;
  di_count: number;
  ao_count: number;
  do_count: number;
  signals_total: number;
};

export type IOTreeCabinet = {
  id: number;
  name: string;
  factory_number?: string | null;
  inventory_number?: string | null;
  channel_devices: IOTreeChannelDevice[];
};

export type IOTreeLocation = {
  id: number;
  name: string;
  children: IOTreeLocation[];
  cabinets: IOTreeCabinet[];
};

export type IOTreeResponse = {
  locations: IOTreeLocation[];
};

export async function fetchIOTree(): Promise<IOTreeResponse> {
  return apiFetch<IOTreeResponse>("/io-tree");
}
